File: backend/app/schemas/labelmaker.py
"""LabelMaker schemas."""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class NutritionValues(BaseModel):
    energyKj: float = 0
    proteinG: float = 0
    fatTotalG: float = 0
    fatSaturatedG: float = 0
    carbohydrateG: float = 0
    sugarsG: float = 0
    sodiumMg: float = 0


class LabelIngredient(BaseModel):
    id: str
    name: str
    allergenTags: list[str] = []
    sulphitesMgPerKg: Optional[float] = None


class LabelTemplateBase(BaseModel):
    prod_product_id: int
    pack_type_code: str = Field(..., max_length=50)
    product_name_zh: str = Field("", max_length=200)
    product_name_en: str = Field(..., max_length=200)
    net_weight_g: Decimal
    serving_size_g: Decimal
    servings_per_package: Decimal = Decimal("1")
    storage_conditions: str
    customer_text: str
    shelf_life_days: int = Field(365, ge=0, le=3650)
    nutrition_per_100g: NutritionValues
    ingredients: list[LabelIngredient]
    recipe: Optional[dict[str, Any]] = None
    allergens_confirmed_at: Optional[datetime] = None


class LabelTemplateCreate(LabelTemplateBase):
    pass


class LabelTemplateUpdate(BaseModel):
    product_name_zh: Optional[str] = Field(None, max_length=200)
    product_name_en: Optional[str] = Field(None, max_length=200)
    net_weight_g: Optional[Decimal] = None
    serving_size_g: Optional[Decimal] = None
    servings_per_package: Optional[Decimal] = None
    storage_conditions: Optional[str] = None
    customer_text: Optional[str] = None
    shelf_life_days: Optional[int] = Field(None, ge=0, le=3650)
    nutrition_per_100g: Optional[NutritionValues] = None
    ingredients: Optional[list[LabelIngredient]] = None
    recipe: Optional[dict[str, Any]] = None
    allergens_confirmed_at: Optional[datetime] = None


class LabelTemplateResponse(LabelTemplateBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_code: Optional[str] = None
    product_name: Optional[str] = None
    pack_type_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class TranslateIngredientRequest(BaseModel):
    query: str = Field(..., max_length=120)


class IngredientTranslationResponse(BaseModel):
    englishTerms: list[str]
    displayHint: str
    confidence: float


class RefineIngredientItem(BaseModel):
    id: str
    fsanzName: str
    currentLabelName: str
    weightG: float
    percentage: float


class RefineIngredientLabelsRequest(BaseModel):
    productName: str = Field("", max_length=160)
    ingredients: list[RefineIngredientItem]


class RefinedIngredientLabel(BaseModel):
    id: str
    labelName: str
    showPercentage: bool


class IngredientLabelRefinementResponse(BaseModel):
    ingredients: list[RefinedIngredientLabel]
    note: str


class LabelPdfRequest(BaseModel):
    template_id: Optional[int] = None
    prod_product_id: Optional[int] = None
    pack_type_code: Optional[str] = None
    production_date: Optional[date] = None
    expiry_date: Optional[date] = None

File: frontend/src/api/labelmaker.ts
import apiClient from './client';
import type { ProductTemplate } from '@/features/labelmaker/types';

export interface LabelTemplate {
  id: number;
  prod_product_id: number;
  pack_type_code: string;
  product_name_zh: string;
  product_name_en: string;
  net_weight_g: string;
  serving_size_g: string;
  servings_per_package: string;
  storage_conditions: string;
  customer_text: string;
  shelf_life_days: number;
  nutrition_per_100g: ProductTemplate['nutritionPer100g'];
  ingredients: ProductTemplate['ingredients'];
  recipe: ProductTemplate['recipe'] | null;
  allergens_confirmed_at: string | null;
  product_code: string | null;
  product_name: string | null;
  pack_type_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface LabelTemplatePayload {
  prod_product_id: number;
  pack_type_code: string;
  product_name_zh: string;
  product_name_en: string;
  net_weight_g: number;
  serving_size_g: number;
  servings_per_package: number;
  storage_conditions: string;
  customer_text: string;
  shelf_life_days: number;
  nutrition_per_100g: ProductTemplate['nutritionPer100g'];
  ingredients: ProductTemplate['ingredients'];
  recipe?: ProductTemplate['recipe'] | null;
  allergens_confirmed_at?: string | null;
}

export const labelmakerApi = {
  listTemplates: async (params?: { prod_product_id?: number; pack_type_code?: string }): Promise<LabelTemplate[]> => {
    const res = await apiClient.get<LabelTemplate[]>('/api/v1/labelmaker/templates', { params });
    return res.data;
  },
  getTemplateByProductPack: async (prodProductId: number, packTypeCode: string): Promise<LabelTemplate> => {
    const res = await apiClient.get<LabelTemplate>('/api/v1/labelmaker/templates/by-product-pack', {
      params: { prod_product_id: prodProductId, pack_type_code: packTypeCode },
    });
    return res.data;
  },
  createTemplate: async (data: LabelTemplatePayload): Promise<LabelTemplate> => {
    const res = await apiClient.post<LabelTemplate>('/api/v1/labelmaker/templates', data);
    return res.data;
  },
  updateTemplate: async (id: number, data: Partial<LabelTemplatePayload>): Promise<LabelTemplate> => {
    const res = await apiClient.patch<LabelTemplate>(`/api/v1/labelmaker/templates/${id}`, data);
    return res.data;
  },
  deleteTemplate: async (id: number): Promise<void> => {
    await apiClient.delete(`/api/v1/labelmaker/templates/${id}`);
  },
  renderPdf: async (data: {
    template_id?: number;
    prod_product_id?: number;
    pack_type_code?: string;
    production_date?: string;
    expiry_date?: string;
  }): Promise<Blob> => {
    const res = await apiClient.post('/api/v1/labelmaker/render-pdf', data, { responseType: 'blob', timeout: 60000 });
    return res.data;
  },
};
